# Exploring Hacker News Post 

In this project, we'll work with a data set of submissions to popular technology site Hacker News.


[Here](https://www.kaggle.com/hacker-news/hacker-news-posts) is the data set that we will work for. Below are descriptions of the columns: 
<ul>
<li> `id`: The unique identifier from Hacker News for the post </li> 
<li> `title`: The title of the post
url: The URL that the posts links to, if it the post has a URL </li>
<li> `num_points`: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
num_comments: The number of comments that were made on the post </li>
<li> `author`: The username of the person who submitted the post </li>
<li> `created_at`: The date and time at which the post was submitted </li> 
</ul> 

We're specifically interested in posts whose titles begin with either `Ask HN` or `Show HN`.
We'll compare these two types of posts to determine the following:
<ul>
<li> Do `Ask HN` or `Show HN` receive more comments on average? </li> 
<li> Do posts created at a certain time receive more comments on average? </li> 
</ul> 

## Introduction
Let's start by importing the libraries we need and reading the data set into a list of lists.

In [7]:
from csv import reader 
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
print(hn[0:5])


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


## Removing Headers from a data set
We display the first five rows of the data set as above and observe that the first row contains the column headers. In order to analyze our data, we need to first remove the row containing the column headers. 

In [10]:
headers = hn[0]
hn = hn[1:]
print(headers)
print("\n")
print(hn[0:5])

['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


[['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12'], ['10482257', 'Title II kills investment? Comcast and other ISPs are now spending more', 'http://arstechnica.com/business/2015/10/comcast-and-other-isps-boost-network-investment-despite-net-neutrality/', '53', '22', 'Deino

## Extracting Ask HN and Show HN Posts
Now that we've removed the headers from `hn`, we're ready to filter our data. Since we're only concerned with post titles beginning with `Ask HN` or `Show HN`, we'll create new lists of lists containing just the data for those titles.

In [18]:
ask_posts = []
show_posts = [] 
other_posts = []
for row in hn:
    title = row[1]
    title_lower = title.lower()
    if title_lower.startswith("ask hn"): 
        ask_posts.append(row)
    elif title_lower.startswith("show hn"): 
        show_posts.append(row)
    else : 
        other_posts.append(row)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
    

1744
1162
17192


We can see that there are 1744 rows start with `Ask HN` and 1162 rows start with `Show HN`. 

## Calculating the Average Number of Comments for Ask HN and Show HN Posts 
Next, let's determine if ask posts or show posts receive more comments on average.

In [20]:
total_ask_comments = 0 
for row in ask_posts: 
    num_com = int(row[4]) 
    total_ask_comments += num_com 
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

total_show_comments = 0 
for row in show_posts: 
    num_com = int(row[4]) 
    total_show_comments += num_com 
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

    

14.038417431192661
10.31669535283993


As we can see, the average number of comments for ask posts is about 14 while the average number of comments for show posts is only more than 10. 
So the ask posts received more comments on average. 

## Finding the Amount of Ask Posts and Comments by Hour Created
Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:
<ul>
<li> Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.</li> 
<li> Calculate the average number of comments ask posts receive by hour created.</li> 

In [29]:
import datetime as dt 
result_list = [] 
for row in ask_posts: 
    result_list.append([row[6],int(row[4])])
counts_by_hour = {} 
comments_by_hour = {} 
for row in result_list: 
    date = row[0]
    date_dt = dt.datetime.strptime(date,"%m/%d/%Y %H:%M")
    hour = date_dt.strftime("%H")
    if hour not in counts_by_hour: 
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
    else: 
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]

## Calculating the Average Number of Comments for Ask HN Posts by Hour
We created two dictionaries as above:
<ul> 
<li> `counts_by_hour`: contains the number of ask posts created during each hour of the day. </li> 
<li> `comments_by_hour` : contains the corresponding number of comments ask posts created at each hour received. </li> 
</ul> 
Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

We will create a list of list in which the first element is the hour and the second element is the average number of comments per post. 

In [31]:
avg_by_hour = [] 
for hour in comments_by_hour: 
    num = counts_by_hour[hour]
    num_com = comments_by_hour[hour]
    avg_by_hour.append([hour, num_com/num])
    
avg_by_hour

[['20', 21.525],
 ['21', 16.009174311926607],
 ['15', 38.5948275862069],
 ['23', 7.985294117647059],
 ['08', 10.25],
 ['12', 9.41095890410959],
 ['18', 13.20183486238532],
 ['16', 16.796296296296298],
 ['02', 23.810344827586206],
 ['13', 14.741176470588234],
 ['14', 13.233644859813085],
 ['01', 11.383333333333333],
 ['07', 7.852941176470588],
 ['19', 10.8],
 ['05', 10.08695652173913],
 ['09', 5.5777777777777775],
 ['22', 6.746478873239437],
 ['17', 11.46],
 ['06', 9.022727272727273],
 ['10', 13.440677966101696],
 ['11', 11.051724137931034],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['03', 7.796296296296297]]

## Sorting and Printing Values from a List of Lists 
Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [42]:
swap_avg_by_hour = [] 
for row in avg_by_hour: 
    swap_avg_by_hour.append([row[1],row[0]])
sorted_swap = sorted(swap_avg_by_hour,reverse = True)
top_five = []
for i in range(5): 
    hour = sorted_swap[i][1]
    hour_dt = dt.datetime.strptime(hour,"%H")
    hour_fm = hour_dt.strftime("%H:%M")
    num = sorted_swap[i][0]
    output = "{hour}: {number:.2f} average comments per post".format(hour = hour_fm, number = num)
    top_five.append(output)
print("Top 5 Hours receive the most comments")
top_five

Top 5 Hours receive the most comments


['15:00: 38.59 average comments per post',
 '02:00: 23.81 average comments per post',
 '20:00: 21.52 average comments per post',
 '16:00: 16.80 average comments per post',
 '21:00: 16.01 average comments per post']

The hour that receives most comments is `15:00` with 38.59 comments on average. 

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est. Converting to my local time, it is 3:00 am GMT+7. 

## Conclusion 
In this project, by analyzing the data set of Hacker News Posts, we observed that there are more comments on average for the ask posts than for the show post. After that, we calculated the amount of ask posts and comments by hour created and found that posts created at 3:00 pm EST (3:00 AM GMT+7) has the highest number of comments. 

So we recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).